### PR TITLE
Add Custom gateway tutorial

### PR DESCRIPTION
**Pull request by TucksonDev**
This tutorial creates a custom gateway on L1 and L2, as well as custom tokens for L1 and L2.
It demonstrates how to use the custom-custom way for bridging tokens.

### CODE REVIEW

File: packages/custom-gateway-bridging/contracts/L1CustomGateway.sol [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    
+    // Token bridge state variables
+    address public l1CustomToken;
+    address public l2CustomToken;
+    address public l2Gateway;
+    address public router;
+    bool private tokenBridgeInformationSet = false;
+
+    // Custom functionality
+    bool public allowsDeposits = false;
```
Comment (gzeoneth) [inline packages/custom-gateway-bridging/contracts/L1CustomGateway.sol:24]: while this is safe bcz false is the default value, setting value when defining storage var can be risky when you use a proxy as it won't be set

File: packages/custom-gateway-bridging/contracts/L1CustomGateway.sol [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    /// @dev See {IL1CustomGateway-outboundTransferCustomRefund}
+    function outboundTransferCustomRefund(
+        address l1Token,
+        address refundTo,
+        address to,
+        uint256 amount,
+        uint256 maxGas,
+        uint256 gasPriceBid,
+        bytes calldata data
+    ) public payable override returns (bytes memory res) {
```
Comment (gzeoneth) [inline packages/custom-gateway-bridging/contracts/L1CustomGateway.sol:81]: you have a named return but ended with `return abi.encode(seqNum);`

File: packages/custom-gateway-bridging/contracts/CrosschainMessenger.sol [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+ * @title Interface needed to call function l2ToL1Sender of the Outbox
+ */
+interface IOutbox {
+    function l2ToL1Sender() external view returns (address);
+}
+
+/**
+ * @title Minimum expected implementation of a crosschain messenger contract to be deployed on L1
+ */
+abstract contract L1CrosschainMessenger {
```
[the remark this answers is not in the document]
Comment (TucksonDev) [reply packages/custom-gateway-bridging/contracts/CrosschainMessenger.sol:42]: token-bridge-contracts is currently in v0.6, and depend on openzeppelin/contracts v3.4.2, making it difficult to integrate with the current implementations.
As discussed, I'll continue with this version and update the tutorial when a new token-bridge-contracts is released.

File: packages/custom-gateway-bridging/contracts/L1CustomGateway.sol (hunk 1 above)
Comment (TucksonDev) [reply packages/custom-gateway-bridging/contracts/L1CustomGateway.sol]: This version is not prepared for a proxy, but in order to minimize the changes to be made to use a proxy, I've moved the default value to the constructor.

**Comment by TucksonDev** [comment]
Thanks @gvladika  and @gzeoneth !
I've updated the code including all your feedback (with the exception of importing the interfaces+abstract contracts from token-bridge-contracts, more info in that comment).
I've also added a test for withdrawals, and another one for the custom functionality (enable/disable deposits)

File: packages/custom-gateway-bridging/contracts/L2CustomGateway.sol [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+        // The inboundEscrowAndCall functionality has been disabled, so no data is allowed
+        require(extraData.length == 0, "EXTRA_DATA_DISABLED");
+
+        // Burns L2 tokens in order to release escrowed L1 tokens
+        IArbToken(l2CustomToken).bridgeBurn(from, amount);
+
+        // Current exit number for this operation, after
+        // incrementing the exit number for future withdrawals
+        uint256 currExitNum = ++exitNum;
```
Comment (gvladika) [inline packages/custom-gateway-bridging/contracts/L2CustomGateway.sol:96]: should be post-increment

**Review by gvladika — APPROVED** [review]
lgtm

**Pull request merged**